rstudio/rstudio
# Source button uses `cpp11::source_cpp()` on C++ files with `[[cpp11::register]]` decorations

### Intent

addresses #10387

### Approach

New rpc method for detecting if the file follows `cpp11` way of using `[[cpp11::register]]` decorations instead of `// [[Rcpp::export]]`. 

When `cpp11` is detected: `cpp11::cpp_source()` is used instead of `Rcpp::sourceCpp()`. 

### Automated Tests

In `test-cpp.R`: 

```r
test_that("cppSourceFile() uses correct package", {
    tf <- tempfile(fileext = ".cpp")
    
    writeLines('#include <Rcpp.h>\n//[[Rcpp::export]]\nvoid fun(){}', tf)
    call <- .rs.cppSourceFile(tf)
    expect_match(call, "Rcpp::sourceCpp")

    writeLines('void fun(){}', tf)
    call <- .rs.cppSourceFile(tf)
    expect_match(call, "Rcpp::sourceCpp")

    writeLines('#include <cpp11.hpp>\n[[cpp11::register]]\nvoid fun(){}', tf)
    call <- .rs.cppSourceFile(tf)
    expect_match(call, "cpp11::cpp_source")
    
    unlink(tf)
    call <- .rs.cppSourceFile(tf)
    expect_match(call, "Rcpp::sourceCpp")
})
```

### QA Notes

- Generate a new cpp11 file with  `File > New File > New C++ file`
- (if needed) replace the code by: 

```cpp
#include <cpp11.hpp>

// This is a simple example of exporting a C++ function to R. You can
// source this function into an R session using `cpp11::cpp_source()`.

// Learn more about cpp11 at: 
//
//   https://cpp11.r-lib.org/index.html

[[cpp11::register]]
double sum(cpp11::doubles x) {
    double total = 0.0;
    for (double value: x) {
        total += value;
    }
    return total;
}
```

- Click the source button
- this should execute `cpp_source(<filename>)` in the console. 

```
> sum(c(1, 2))
[1] 3
```

### Checklist
